File: LeetCode/Java Solutions/Boats to Save People/boats.java
import java.util.Arrays;

class Solution {
    public int numRescueBoats(int[] people, int limit) {
        Arrays.sort(people);
        int left = 0, right = people.length - 1;
        int boats = 0;
        
        while (left <= right) {
            if (people[left] + people[right] <= limit) {
                left++;  // Pair the lightest and heaviest person that can fit together
            }
            right--;  // Heaviest always gets on a boat, either alone or paired
            boats++;
        }
        
        return boats;
    }
}

File: LeetCode/Python Solutions/Boats to Save People/boats.py
from typing import List

class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()
        left, right = 0, len(people) - 1
        boats = 0
        
        while left <= right:
            if people[left] + people[right] <= limit:
                left += 1  # The lightest person is paired
            # The heaviest person always goes on the boat
            right -= 1
            boats += 1
        
        return boats
